docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,509 @@
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Abstract
+
+Real Kubernetes clusters have a variety of volumes which differ widely in
+size, iops performance, retention policy, and other characteristics.
+Administrators need a way to dynamically provision volumes of these different
+types to automatically meet user demand.
+
+A new mechanism called 'storage classes' is proposed to provide this
+capability.  Storage classes build upon the label selection feature added in
+Kubernetes 1.3.
+
+## Motivation
+
+In Kubernetes 1.2, an alpha form of limited dynamic provisioning was added
+that allows a single volume type to be provisioned in clouds that offer
+special volume types.
+
+In Kubernetes 1.3, a label selector was added to persistent volume claims to
+allow administrators to create a taxonomy of volumes based on the
+characteristics important to them, and to allow users to make claims on those
+volumes based on those characteristics.  This allows flexibility when claiming
+existing volumes; the same flexibility is needed when dynamically provisioning
+volumes.
+
+After gaining experience with dynamic provisioning after the 1.2 release, we
+want to create a more flexible feature that allows configuration of how
+different storage classes are provisioned and supports provisioning multiple
+types of volumes within a single cloud.
+
+### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Design
+
+We propose that:
+
+1.  Storage properties of peristent volumes is represented as labels.
+    E.g. label "aws.alpha.kubernetes.io/volume-type" with values "sc1" and "gp2"
+    represent SDD/HDD of AWS EBS. Similarly, label
+    "failure-domain.beta.kubernetes.io/region" with values like "us-east-1d"
+    represent zone where particular AWS EBS is created.
+
+    Users will request specific storage volumes by specifying
+    `claim.Spec.Selector`, such as `region NotIn [us-east-1a, us-east-1b] AND
+    volume-type == sc1`
```

Comment:
Yesterday I think we discussed making the usage of `claim.Spec.Selector` and `claim.Spec.Class` (or the annoation) mutually exclusive and objects with both set would fail validation. We would see what people thought of that.

We could come back in the future and enable the usage of BOTH `selector` and `class` at the same time...